daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/TDMLRunner.scala

```diff
@@ -1806,68 +1881,101 @@ object VerifyTestCase {
    * @param expectedDiags                             Expected diagnostics from test cases
    * @param implString                                Implementation string
    */
-  def verifyAllDiagnosticsFound(
+  def verifyDiagnosticsFound(
     actualDiags: Seq[Diagnostic],
-    expectedDiags: Option[ErrorWarningBase],
+    expectedDiags: Option[Seq[ErrorWarningBase]],
+    ignoreUnexpectedWarnings: Boolean,
     implString: Option[String]
-  ) = {
+  ): Option[Seq[Unit]] = {
 
-    val actualDiagMsgs = actualDiags.map {
-      _.toString()
-    }
     val expectedDiagMsgs = expectedDiags
-      .map {
-        _.messages
+      .map { d =>
+        d.flatMap(_.messages)
       }
       .getOrElse(Nil)
 
-    if (expectedDiags.isDefined && actualDiags.isEmpty) {
+    val expectedDiagnosticType = if (expectedDiags.isDefined) {
+      expectedDiags.get.head.diagnosticType
+    } else {
+      ""
+    }
+    val actualDiagsFiltered = if (expectedDiagnosticType == "validation error") {
+      actualDiags.filter(diag => diag.isValidation)
+    } else if (expectedDiagnosticType == "warning") {
+      actualDiags.filterNot(d => d.isValidation || d.isError)
+    } else if (expectedDiagnosticType == "error") {
+      actualDiags.filter(_.isError)
+    } else {
+      Seq.empty
+    }
```

Comment:
I general prefer to avoid comparing strings, they tend to be fragile. For example, if we one day decide to capitalize these this stuff will break. Instead, we can use Scala type system, e.g.

```
val expectedDiagType = expectedDiags.headOption
val actualDiagsFiltered = expectedDiagType match {
  case Some(_: ExpectedValidationError) => actualDiags.filter(diag => diag.isValidation)
  case Some(_: ExpectedWarning) => actualDiags.filterNot(d => d.isValidation || d.isError)
  case Some(_: ExpectedError) => actualDiags.filter(_.isError)
  case None => Seq.empty
}
```
That said, it might be useful to pass in an enum or a filter function or something as a paramter, or do the filtering in checkDiagnosticMessages (assuming that can be updated so everyone calls it). Something like that might be preferred, since this ends up with `Seq.empty` if we don't know the type, which makes it look like there are no actual diagnostics to check.